source/MRVoxels/MRWeightedPointsShell.cpp

```diff
@@ -149,11 +154,72 @@ Expected<Mesh> weightedMeshShell( const Mesh & mesh, const WeightedPointsShellPa
     return marchingCubes( weightedMeshToDistanceFunctionVolume( mesh, wp2vparams ), vmParams );
 }
 
-Expected<Mesh> weightedMeshShell( const Mesh & mesh, const VertScalars& vertWeights, const WeightedPointsShellParameters& params0 )
+Expected<Mesh> weightedMeshShell( const Mesh & mesh, const VertScalars& vertWeights, const WeightedPointsShellParametersMetric& params0 )
 {
     auto params = params0;
     params.dist.pointWeight = [&vertWeights]( VertId v ){ return vertWeights[v]; };
     return weightedMeshShell( mesh, params );
 }
 
+Expected<Mesh> weightedMeshShell( const Mesh& mesh, const WeightedPointsShellParametersRegions& params )
+{
+    MR_TIMER
+
+    if ( params.regions.empty() )
+        spdlog::warn( "weightedMeshShell called without regions. Consider using MR::offsetMesh which is more efficient for constant offset." );
+
+    VertBitSet allVerts;
+    for ( const auto& reg : params.regions )
+        allVerts |= reg.verts;
+
+    const float interRadSq = sqr( params.interpolationDist );
+    auto pointWeight = [&params, &mesh, &allVerts, interRadSq] ( VertId v )
+    {
+        if ( params.regions.empty() )
+            return 0.f;
+        if ( !allVerts.test( v ) )
+            return 0.f;
+        float res = 0.f;
+        size_t n = 0;
+
+        const auto pt = mesh.points[v];
+        findPointsInBall( mesh, Ball3f{ pt, interRadSq }, [&n, &res, &params]
+            ( const PointsProjectionResult & found, const Vector3f &, Ball3f & )
+        {
+            auto vv = found.vId;
+            for ( const auto& reg : params.regions )
+            {
+                if ( reg.verts.test( vv ) )
+                {
+                    res += reg.weight;
+                    n += 1;
+                }
+            }
+            return Processing::Continue;
+        } );
+
+        assert( n > 0 );
+        return res / static_cast<float>( n );
+    };
+
+    // precalculate the weights
+    VertScalars weights( mesh.topology.getValidVerts().find_last() + 1, 0 );
+    BitSetParallelFor( allVerts, [&weights, &pointWeight] ( VertId v )
+    {
+        weights[v] = pointWeight( v );
+    } );
+
+    DistanceFromWeightedPointsParams distParams;
+    distParams.maxWeight = parallelMinMax( weights, &allVerts ).second;
```

Comment:
Is it right that maxWeight for points is the max weight of regions? Then it can be computed much fastrer.